### PR TITLE
Handle Files upload.

### PR DESCRIPTION
**Pull request by mathieutu**
Hi,
With this trick, we automatically convert form data to a Javascript FormData object, which handles natively File uploads (and request weight is much lighter by this way than with data-url).

Some tests should be added, but as I never done javascripts tests, I may need some help. 

(and I removed cloneDeep import which was never used anymore)

What are your thoughts about that guys?

See you,
Matt'

### CODE REVIEW

**Comment by freekmurze** [comment]
Seems nice, going to let @sebastiandedeyne decide on this one.

Could you already rewrite the code so you don't make use of any else statements? Introduce some more helper functions so you can use early returns. 

Also is that `hasFiles` really needed? Couldn't we just do the processing every time?

**Comment by mathieutu** [comment]
> Could you already rewrite the code so you don't make use of any else statements? Introduce some more helper functions so you can use early returns.

No problem for that, will do that as I get a bit of time. It could be nice if I know before if you want to merge the PR, it can avoid useless work. ;-)

>Also is that hasFiles really needed? Couldn't we just do the processing every time?

As you want, imo we should keep it, as it avoids some work (creating the FormData and appending all the properties) if there is no need, and it leaves the request in json (standard way?) if there has no file.

**Comment by mathieutu** [comment]
Hi, just to tell you that I'm still alive, I just didn't have the time to work on this previously, and I'm in vacations for now!

I've planed to work on this when I'll be back from the mountain 😛.

**Comment by sebastiandedeyne** [comment]
Don't sweat it, enjoy your holiday! :D

**Comment by mathieutu** [comment]
Hi!
As promised, I'm back! 
I've tried to refactor the helpers in a proper way but keep them simple to read.
I also have added some tests (for file upload and some others) and set Travis to launch them. 

For the readme, I don't really know what to write, and as (Freek does know that) my english isn't really that good, I prefer let you add something if you want. But, it really works out of the box, you don't need to think about it. Just set the field(s) with the File(s), and it's done.

For vuejs, it can be easier : 
```html
<input type="file" @change="form.file = $event.target.files[0]"/> 
```

Let me know if I can do more,

See you!
Matt'

File: __tests__/Form.test.js [hunk 1]
```diff
@@ -208,4 +206,44 @@ describe('Errors', () => {
...
+    it('can call directly HTTP verbs to submit', () => {
+        form.submit = (...args) => args;
+
+        expect(form.post('url')).toEqual(['post', 'url']);
+        expect(form.put('url')).toEqual(['put', 'url']);
+        expect(form.patch('url')).toEqual(['patch', 'url']);
+        expect(form.delete('url')).toEqual(['delete', 'url']);
+    });
+
+    it('transforms the data ta a FormData object if there is a File', async () => {
```
Comment (ankurk91) [inline __tests__/Form.test.js]: @mathieutu

> transforms the data **ta** a


There is a typo in this

Comment (mathieutu) [reply __tests__/Form.test.js]: Yeah, thanks!

**Comment by sebastiandedeyne** [comment]
Looks great!

Gonna merge this in and do a few more little personal refactors, thanks!

**Comment by mathieutu** [comment]
ok, I'll pay attention to them. Don't miss the typo! ;-)

**Comment by M-Abdullahi** [comment]
Thank you @mathieutu  and the rest for implementing  this feature.

**Pull request merged**